### ISSUE TITLE
Sudden slow down at datastore queries

### ISSUE THREAD

**Issue by mashbrno** [canonical/microk8s#2205]
Cluster of 4 nodes was running fine for several months. Today we observed an issue with very slow pod rescheduling. Creating a *pod* directly is still matter of seconds, but creating a *deployment* is taking 10 minutes or more.

The log is full of similar very long queries:

```
"List etcd3" key:/acme.cert-manager.io/orders,resourceVersion:,resourceVersionMatch:,limit:10000,continue: (26-Apr-2021 13:29:56.063) (total time: 30728ms): 
```

Which we believe are root cause of the whole system unresponsiveness. There appears timeouts in various parts of the system, `kubectl` commands are not responding.

[inspection-report-20210426_160422.tar.gz](https://github.com/ubuntu/microk8s/files/6379807/inspection-report-20210426_160422.tar.gz)

**Comment by balchua** [comment]
I think the `kube-controller-manager` is crashlooping.  I see these in the logs.

```
Apr 26 15:54:06 microk8s3 microk8s.daemon-controller-manager[695948]: E0426 15:54:06.803372  695948 leaderelection.go:361] Failed to update lock: Put "https://127.0.0.1:16443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-controller-manager?timeout=10s": context deadline exceeded
```
The leader election is failing, preventing it to function properly.  As a mitigation, you can bump these 2 arguments
```
--leader-elect-lease-duration=60s
--leader-elect-renew-deadline=60s
```
Into these files:
`/var/snap/microk8s/current/args/kube-controller-manager` and `/var/snap/microk8s/current/args/kube-scheduler`

Then restart MicroK8s.

There is also an ongoing PR to mitigate such behavior.
https://github.com/ubuntu/microk8s/pull/2203

**Comment by mashbrno** [comment]
Thanks for updating the original comment, renew interval has to be shorter than lease due to `couldn't create leader elector: leaseDuration must be greater than renewDeadline`.

I have restarted the cluster and the error messages are not appearing now.

**Comment by mashbrno** [comment]
Just for further reference on 25th April there was new 1.20.6 release of microk8s snap, which was probably installed during early morning on 26th - the day when the problem was observed. Are the snap upgrades coordinated through whole cluster?

**Comment by mashbrno** [comment]
Ended with:

```
microk8s status
microk8s is running
high-availability: no
  datastore master nodes: none
  datastore standby nodes: none
```

**Comment by ktsakalozos** [comment]
> Are the snap upgrades coordinated through the whole cluster?

Snap upgrades will happen when snapd refreshes the snap. You have control over when such refreshes will occur, have a look at the end of https://microk8s.io/docs/setting-snap-channel .



> Ended with ....

No master nodes no standby ones? Is the cluster running? Can you share a `microk8s.inspect` tarball?

**Comment by MathieuBordere** [comment]
@mashbrno Is your cluster operational at the moment? Do you still observe the slowdown?

**Comment by mashbrno** [comment]
@ktsakalozos In fact you confirm my hypothesis that if snap is left in default configuration it can happen that multiple nodes refreshes (and updates) simultanously which may break my pretty small 4 node cluster.

The downtime was from 14:40 to 16:00 local time. The nodes were going down one by one ended with no masters no standby nodes. I tried to upgrade a node to 1.21/stable as there should be some improvements in dqlite, but the node hadn't started in reasonable time, so I reverted the snap. Instead of I tried to join a completely new node which was failing (500) for some nodes, but finally I have found one to join to.

[inspection-report-20210428_194714.tar.gz](https://github.com/ubuntu/microk8s/files/6393893/inspection-report-20210428_194714.tar.gz)
